src/kademlia/node.cpp

```diff
@@ -453,7 +452,77 @@ void node::get_item(char const* pk, std::string const& salt
 #endif
 
 	boost::intrusive_ptr<dht::get_item> ta;
-	ta.reset(new dht::get_item(*this, pk, salt, f));
+	ta.reset(new dht::get_item(*this, pk, salt, f, find_data::nodes_callback()));
+	ta->start();
+}
+
+namespace {
+
+void put(std::vector<std::pair<node_entry, std::string> > const& nodes
+	, boost::intrusive_ptr<dht::put_data> ta)
+{
+    ta->set_targets(nodes);
+	ta->start();
+}
+
+void put_data_cb(item& i, bool auth
+		, boost::intrusive_ptr<put_data> ta
+		, boost::function<void(item&)> f)
+{
+	// call data_callback only when we got authoritative data.
```

Comment:
this comment makes me suspicious though. The point of the authoritative flag is for the user to get multiple data callbacks, the first ones coming back very quickly, but without being authoritative and then the last one (which probably has to wait for a few node time-outs, and will be very late) being authoritative. i.e. not until the last one can one be somewhat certain that it is the latest version.

The benefit of this is that clients can start perform work optimistically, and discard it later if it turned out it wasn't the latest version of the data. But chances are that the first data blob you get back _is_ the latest, or at least contain some relevant data.